server/project_member.go

```diff
@@ -2,18 +2,146 @@ package server
+		}
+
+		projectFind := &api.ProjectFind{ID: &projectID}
+		project, err := s.ProjectService.FindProject(ctx, projectFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Project not found: %s", c.Param("projectID"))).SetInternal(err)
+		}
+		if project.WorkflowType == api.UIWorkflow {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid workflow type: %s, need %s to enable this function", project.WorkflowType, api.VCSWorkflow))
+		}
+
+		repoFind := &api.RepositoryFind{ProjectID: &projectID}
+		repo, err := s.RepositoryService.FindRepository(ctx, repoFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch relevant VCS repo, Project ID: %s", c.Param("projectID"))).SetInternal(err)
+		}
+
+		vcsFind := &api.VCSFind{
+			ID: &repo.VCSID,
+		}
+		vcs, err := s.VCSService.FindVCS(ctx, vcsFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find VCS for sync project member: %d", repo.VCSID)).SetInternal(err)
+		}
+		if vcs == nil {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("VCS ID not found: %d", repo.VCSID))
+		}
+
+		vcsProjectMemberList, err := vcsPlugin.Get(vcs.Type, vcsPlugin.ProviderConfig{Logger: s.l}).FetchRepositoryActiveMemberList(ctx,
+			common.OauthContext{
+				ClientID:     vcs.ApplicationID,
+				ClientSecret: vcs.Secret,
+				AccessToken:  repo.AccessToken,
+				RefreshToken: repo.RefreshToken,
+				Refresher:    s.refreshToken(ctx, repo.ID),
+			},
+			vcs.InstanceURL,
+			repo.ExternalID,
+		)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch repository member from VCS, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+		}
+
+		findProjectMember := &api.ProjectMemberFind{ProjectID: &projectID}
+		bytebaseProjectMemberList, err := s.ProjectMemberService.FindProjectMemberList(ctx, findProjectMember)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch repository member in bytebase: %d", projectID)).SetInternal(err)
+		}
+
+		// check whether principal exists in our system.
+		// If not exist, create one.
+		for _, projectMember := range vcsProjectMemberList {
+			findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+			principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+			}
+
+			isPrincipalNewCreated := false
+			if principal == nil {
+				signupInfo := &api.Signup{
+					Name:  projectMember.Name,
+					Email: projectMember.Email,
+					// Principal created via this method would have not change to set their password,
+					// To prevent potential safe issue, we use random string to set up her password.
```

Comment:
```suggestion
					// Principal created via this method would have not chance to set their password,
					// To prevent potential safe issue, we use random string to set up her password.
					// This is another safety measure since we already disallow user login via password
					// if the principal uses external auth provider
```